packages/compass-aggregations/src/modules/pipeline-builder/pipeline-parser/pipeline-parser.ts

```diff
@@ -0,0 +1,189 @@
+    return !trailingVisited.has(node);
+  });
+  return [...leadingStages, node, ...trailingStages];
+}
+
+type ParseResponse = {
+  root: t.ArrayExpression;
+  stages: t.Expression[];
+};
+
+type ValidateResponse = {
+  root: t.ArrayExpression | null;
+  errors: SyntaxError[];
+};
+
+export default class PipelineParser {
+  static parse(source: string): ParseResponse {
+    const stages: t.Expression[] = [];
+    const root = PipelineParser._parseStringToRoot(source);
+    // Inner comments will only be available here if there is no other
+    // elements in the array, in our case it might mean that all stages
+    // are disabled
+    if (root.elements.length === 0 && root.innerComments?.length) {
+      const [, _stages] = extractStagesFromComments(
+        root.innerComments
+      );
+      stages.push(..._stages);
+    }
+    root.elements.forEach((node) => {
+      stages.push(...extractStagesFromNode(node as t.Expression));
+    });
+    return { root, stages };
+  }
+
+  static _parseStringToRoot(source: string): t.ArrayExpression {
+    const root = babelParser.parseExpression(source);
+    if (root.type !== 'ArrayExpression') {
+      throw new SyntaxError('Pipeline must be an array of aggregation stages');
+    }
+    return root;
+  }
+
+  // Validate that source is parseable and all stages look like valid pipeline
+  // stages
+  static validate(source: string): ValidateResponse {
+    let root: t.ArrayExpression | null = null;
+    const errors: SyntaxError[] = [];
+    try {
+      root = PipelineParser._parseStringToRoot(source);
+      root.elements.forEach(stage => {
+        try {
+          assertStageNode(stage as t.Expression);
+        } catch (e) {
+          errors.push(e as SyntaxError);
+        }
+      });
+    } catch (e) {
+      errors.push(e as SyntaxError);
+    }
+    return { root, errors };
+  }
+  // Generate source from stages
+  static generate(root: t.ArrayExpression, stages: t.Expression[]): string {
+    const isAllDisabled = stages.length && stages.every((stage) => {
+      return isNodeDisabled(stage);
+    });
+    // Special case where all stages should be added as inner comments to the
+    // array expression
+    if (isAllDisabled) {
+      root.elements = [];
+      root.innerComments = stages.flatMap((stage) => {
+        return stageToComments(stage);
+      });
+    } else {
+      root.elements = PipelineParser._getStageNodes(stages);
+    }
+    return generate(root);
+  }
+
+  static _getStageNodes(stages: t.Expression[]): t.ArrayExpression['elements'] {
+    const elements: t.ArrayExpression['elements'] = [];
+    let unusedComments: t.CommentLine[] = [];
+
+    for (const stage of stages) {
+      if (!isNodeDisabled(stage)) {
+        elements.push(stage);
+        continue;
+      }
+
+      const comments = stageToComments(stage);
+      const prevStage = elements[elements.length - 1];
+      if (!prevStage) {
+        unusedComments = unusedComments.concat(comments);
```

Comment:
Not that we really need to microoptimise here, but I thought it would be a good thing to keep in mind: as these `unusedComments` never leave this function you can push in existing array here, instead of creating a new one every time you add an item